### ISSUE TITLE
efibootmgr sets wrong EFI paths when adding boot entries on an xps 9350 

### ISSUE THREAD

**Issue by whirm** [rhinstaller/efibootmgr#47]
```
BootCurrent: 0001
Timeout: 0 seconds
BootOrder: 0003,0000,0001
Boot0000* dobian	PciRoot(0x0)/Pci(0x17,0x0)/Sata(3,32768,0)/HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
Boot0001* debian	PciRoot(0x0)/Pci(0x17,0x0)/Sata(2,65535,0)/HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
Boot0003* dXbian	PciRoot(0x0)/Pci(0x17,0x0)/Sata(3,32768,0)/HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
```

Only 0001 will boot.

0001 is created from the laptop's UEFI setup, 0000 is created using 0.12-4 from Debian and 0003 with current git (5e9700c2252eed45f4568f3a7c08c866c2c83c0b).

The command used in all cases was:

```sh
sudo ./efibootmgr -v -c -e 3 -L dXbian -l '\EFI\debian\grubx64.efi' 
```
Note that I'm using -e 3 as otherwise the boot entries won't even show on the boot menu (efibootmgr will show them allright)

The laptop's BIOS is up to date.

Let me know what other info I can provide.

**Comment by whirm** [comment]
Quick update: I've upgraded to the recently released  BIOS version and and to  kernel git HEAD from a few days ago. 

Now the generated entry looks like this:
```
Boot0000* d3bian	HD(1,GPT,d59c57e1-3e52-4200-bf89-929a0ab0e6d1,0x800,0x800)/File(\EFI\debian\grubx64.efi)
```

**Comment by whirm** [comment]
Another update: depending on if I boot the laptop with a USB disk plugged in or not, the above command will generate boot entries starting with `HD(...` or  `PciRoot(...`. So it has nothing to do with upgrading the BIOS.

**Comment by superm1** [comment]
Are you using an NVMe disk?  It might be worth testing efivar 0.24 + efibootmgr master together to see if the entries created by default are wrong still if so.

Also, are you sure you need the ```-e 3```?  I'm surprising that an entry shouldn't show up in the BIOS without it.

Can you remove all entries, and create one without ```-e 3``` share the entry and try again?

**Comment by whirm** [comment]
I'm using the "SAMSUNG SSD CM871 M.2 2280 128GB" (m2 sata disk).

I've tried building latest git HEAD with:
```sh
rm -fR inst
cd efivar
git pull
make clean
make -j${CONCURRENCY_LEVEL}
DESTDIR=$PWD/../inst make install
cd ../efibootmgr
git pull
make clean
CFLAGS="-I$PWD/../inst/usr/include/efivar" make -j${CONCURRENCY_LEVEL}
DESTDIR=$PWD/../inst make install
```

But efibootdump doesn't compile anymore:

```
make[2]: Leaving directory '/var/data/src/efi/efibootmgr/src'
gcc -I/home/whirm/src/efi/efibootmgr/../inst/usr/include/efivar -I/var/data/src/efi/efibootmgr/src/include -Werror -Wall -Wextra -Wsign-compare -Wstrict-aliasing -std=gnu11 -fshort-wchar -fPIC -flto -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -DLOCALEDIR=\"/usr/share/locale/\"   -I/usr/include/efivar  -c -o efibootmgr.o efibootmgr.c
gcc -I/home/whirm/src/efi/efibootmgr/../inst/usr/include/efivar -I/var/data/src/efi/efibootmgr/src/include -Werror -Wall -Wextra -Wsign-compare -Wstrict-aliasing -std=gnu11 -fshort-wchar -fPIC -flto -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -DLOCALEDIR=\"/usr/share/locale/\"   -I/usr/include/efivar  -c -o efi.o efi.c
gcc -I/home/whirm/src/efi/efibootmgr/../inst/usr/include/efivar -I/var/data/src/efi/efibootmgr/src/include -Werror -Wall -Wextra -Wsign-compare -Wstrict-aliasing -std=gnu11 -fshort-wchar -fPIC -flto -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -DLOCALEDIR=\"/usr/share/locale/\"   -I/usr/include/efivar  -c -o unparse_path.o unparse_path.c
gcc -I/home/whirm/src/efi/efibootmgr/../inst/usr/include/efivar -I/var/data/src/efi/efibootmgr/src/include -Werror -Wall -Wextra -Wsign-compare -Wstrict-aliasing -std=gnu11 -fshort-wchar -fPIC -flto -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -DLOCALEDIR=\"/usr/share/locale/\"   -I/usr/include/efivar    -fno-merge-constants -Wl,--fatal-warnings,--no-allow-shlib-undefined,--default-symver -Wl,-O2 -Wl,--no-undefined-version   -o efibootmgr efibootmgr.o efi.o unparse_path.o  -lefivar -lefiboot
gcc -I/home/whirm/src/efi/efibootmgr/../inst/usr/include/efivar -I/var/data/src/efi/efibootmgr/src/include -Werror -Wall -Wextra -Wsign-compare -Wstrict-aliasing -std=gnu11 -fshort-wchar -fPIC -flto -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -DLOCALEDIR=\"/usr/share/locale/\"   -I/usr/include/efivar  -c -o efibootdump.o efibootdump.c
gcc -I/home/whirm/src/efi/efibootmgr/../inst/usr/include/efivar -I/var/data/src/efi/efibootmgr/src/include -Werror -Wall -Wextra -Wsign-compare -Wstrict-aliasing -std=gnu11 -fshort-wchar -fPIC -flto -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -DLOCALEDIR=\"/usr/share/locale/\"   -I/usr/include/efivar    -fno-merge-constants -Wl,--fatal-warnings,--no-allow-shlib-undefined,--default-symver -Wl,-O2 -Wl,--no-undefined-version   -o efibootdump efibootdump.o unparse_path.o  -lefivar -lefiboot -lpopt
/tmp/cccTkF4o.ltrans0.ltrans.o: In function `main':
<artificial>:(.text+0xb84): undefined reference to `efi_guid_cmp'
collect2: error: ld returned 1 exit status
/var/data/src/efi/efibootmgr/Make.rules:21: recipe for target 'efibootdump' failed
make[1]: *** [efibootdump] Error 1
make[1]: Leaving directory '/var/data/src/efi/efibootmgr/src'
Makefile:10: recipe for target 'all' failed
make: *** [all] Error 2
```

efibootmgr still compiles,


```
┌─(~)───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(whirm@1k)─
└──> sudo efibootmgr -v
BootCurrent: 0001
Timeout: 0 seconds
BootOrder: 0001
Boot0001* debian    PciRoot(0x0)/Pci(0x17,0x0)/Sata(2,65535,0)/HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
┌─(~)───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(whirm@1k)─
└──> cd src/efi/efibootmgr/src/include
┌─(~/src/efi/efibootmgr/src/include)────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(whirm@1k)─
└──> cd ..
┌─(~/src/efi/efibootmgr/src)────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(whirm@1k)─
└──> ls
efibootdump.8  efibootdump.c  efibootdump.o  efibootmgr  efibootmgr.8  efibootmgr.c  efibootmgr.o  efibootnext.c  efi.c  eficonman.c  efi.o  include  Makefile  unparse_path.c  unparse_path.o
┌─(~/src/efi/efibootmgr/src)────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(whirm@1k)─
└──> ../
┌─(~/src/efi/efibootmgr/src)────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(whirm@1k)─
└──> sudo ./efibootmgr -v -c -e 3 -L debian_e3 -l '\EFI\debian\grubx64.efi'
BootCurrent: 0001
Timeout: 0 seconds
BootOrder: 0000,0001
Boot0001* debian    PciRoot(0x0)/Pci(0x17,0x0)/Sata(2,65535,0)/HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
Boot0000* debian_e3 PciRoot(0x0)/Pci(0x17,0x0)/Sata(3,32768,0)/HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
┌─(~/src/efi/efibootmgr/src)────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(whirm@1k)─
└──> sudo ./efibootmgr -v -c  -L debian_no_e3 -l '\EFI\debian\grubx64.efi'
BootCurrent: 0001
Timeout: 0 seconds
BootOrder: 0002,0000,0001
Boot0000* debian_e3 PciRoot(0x0)/Pci(0x17,0x0)/Sata(3,32768,0)/HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
Boot0001* debian    PciRoot(0x0)/Pci(0x17,0x0)/Sata(2,65535,0)/HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
Boot0002* debian_no_e3  HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
┌─(~/src/efi/efibootmgr/src)────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(whirm@1k)─
└──> sudo ./efibootmgr -v -c -e 1 -L debian_e1 -l '\EFI\debian\grubx64.efi'
BootCurrent: 0001
Timeout: 0 seconds
BootOrder: 0003,0002,0000,0001
Boot0000* debian_e3 PciRoot(0x0)/Pci(0x17,0x0)/Sata(3,32768,0)/HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
Boot0001* debian    PciRoot(0x0)/Pci(0x17,0x0)/Sata(2,65535,0)/HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
Boot0002* debian_no_e3  HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
Boot0003* debian_e1 EDD10(0x80)/HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
```


The `debian` entry (created directly from UEFI setup) is the only entry
that works.

I've taken pictures of how each entry shows from the UEFI setup.

Note that debian_no_e3 doesn't even show.
![img_20160630_223317](https://cloud.githubusercontent.com/assets/529494/16514682/c1a39912-3f6d-11e6-88d1-b699e3b2d2f9.jpg)
![img_20160630_223343](https://cloud.githubusercontent.com/assets/529494/16514685/c1a794ea-3f6d-11e6-9712-c5085dd54118.jpg)
![img_20160630_223407](https://cloud.githubusercontent.com/assets/529494/16514686/c1ab1994-3f6d-11e6-8ece-0dae5b28ccaf.jpg)
![img_20160630_223423](https://cloud.githubusercontent.com/assets/529494/16514683/c1a413e2-3f6d-11e6-937b-bf4302743a68.jpg)
![img_20160630_223440](https://cloud.githubusercontent.com/assets/529494/16514684/c1a61c82-3f6d-11e6-9449-a3c7df38514f.jpg)

**Comment by whirm** [comment]
Let me know if I can run any more tests.

**Comment by superm1** [comment]
Alright so a few things.
1. That's really odd the compile failure even against efivar 0.24.  @vathpela will probably need to double check that.
2. Can you make sure you are on latest BIOS release?  For the 9350 it's 1.3.3 currently: https://secure-lvfs.rhcloud.com/downloads/d3afa577b500700365dcdaaef8a573338f4f044b-firmware.cab
3. Can you try to remove ALL entries and only create the one entry without EDD pointing (like this attempt: ```sudo ./efibootmgr -v -c  -L debian_no_e3 -l '\EFI\debian\grubx64.efi'```)
I believe there is a set of code in the BIOS that may be confused by multiple entries pointing to the same thing that you might be hitting.

**Comment by superm1** [comment]
I'm pretty sure you don't have efivar 0.24 installed, I just did a test compile with efivar 0.24 and efibootmgr master does build against it properly.

If it's easier for you to work with, I have efivar's 0.24 package already staged on Alioth:
http://anonscm.debian.org/cgit/uefi/efivar.git/ (branch debian).

It's just waiting on the next efibootmgr release to be pushed into unstable.  You can grab efibootmgr's debian-next branch if you want packages that build against it (http://anonscm.debian.org/cgit/uefi/efibootmgr.git/)

**Comment by whirm** [comment]
I'm running 1.4.3, before that I was running 1.3.3 and it failed too.
```
Libsmbios version:      2.3.0
Product Name:           XPS 13 9350
Vendor:                 Dell Inc.
BIOS Version:           1.4.3
```

I've deleted all of them and it worked!

Is `-e 3` buggy on efibootmgr then? It points to a different SATA device than when adding the same file via UEFI setup and I guess that's why it shows on the boot menu.  

I've also tried adding another entry with 
```
efibootmgr   -c      -L "Debian (EFI stub)" -l '\EFI\debian\vmlinuz' -u "root=UUID=$UUID rootfstype=$TYPE $GRUB_CMDLINE_LINUX_DEFAULT add_efi_memmap initrd=\\EFI\\debian\\initrd.img"
```

And it found the kernel but didn't load the initrd (panicked when it tried to mount the root fs). But I guess this is me not doing  something right or a different bug altogether. 
The entry generated by this last command didn't even work for me with the previous BIOS version and now both efibootmgr from Debian and from git work just fine.

Thanks a lot for your help!

**Comment by whirm** [comment]
> I'm pretty sure you don't have efivar 0.24 installed, I just did a test compile with efivar 0.24 and efibootmgr master does build against it properly.

You are right, It was attempting to use the system installed headers instead of the ones I was pointing it to. I'll try to fix that and come back to you.

**Comment by whirm** [comment]
Yup it's building just fine!

**Comment by vathpela** [comment]
FWIW, in this case -e 3 is probably working because it causes it to put the full ACPI path into the device path instead of the abbreviated HD() path - though I'm not quite sure why that would work, but it's the biggest difference -e 3 will actually cause.

But did 0.24 + efibootmgr HEAD work better?

**Comment by whirm** [comment]
@vathpela `-e 3` is not working (not even with latest git versions) it sets the wrong path (see pics `...sata(0x2,0x0)...` vs `...sata(0x3,0x0)...`)

I haven't appreciated any difference in behavior re 0.24 vs HEAD: both of them will make EFI entries that work if there's only one of them pointing to a given file. (BIOS bug) and both of them will create bad entries if passing `-e 3` as an argument. (everything looks fine except it points to `sata(3...` instead of  `sata(2....`)

I also found out that no arguments get passed to the efi image (I guess it's a BIOS bug too)

**Comment by superm1** [comment]
@whirm 
Ah OK, 1.4.3 is fine too.

> Is -e 3 buggy on efibootmgr then? It points to a different SATA device than when adding the same file via UEFI setup and I guess that's why it shows on the boot menu.

I'd say that's the reason it showed up on the BIOS boot menu yes.  In the general scenario the behavior that the BIOS does to not show duplicate entries in the BIOS boot menu is intended behavior.

> And it found the kernel but didn't load the initrd (panicked when it tried to mount the root fs). But I guess this is me not doing something right or a different bug altogether. 

The lack of arguments getting passed to the EFI image - I think you are building your command line list incorrectly.  When the initramfs is loaded I think the slashes need to face the other way and not be escaped.  The kernel is going to use this to find the initrd, not UEFI.

So I think of the stuff you've found the only efibootmgr issue is that using EDD 3.0 isn't selecting the correct device for you.

**Comment by whirm** [comment]
> The lack of arguments getting passed to the EFI image - I think you are building your command line list incorrectly. When the initramfs is loaded I think the slashes need to face the other way and not be escaped. The kernel is going to use this to find the initrd, not UEFI.

I've build a kernel that doesn't need initrd. And has the default arguments to `root=sda3` so it can always boot if no arguments are received.
Then I've tried creating entries with back and forward slashes for the initrd, with and without having the whole set of arguments wrapped in quotes, with and without  `-u` and the kernel never receives any argument.

I've also tried with `0.24` and `HEAD` just in case.

Some of the attempts I made:
```
+ cd /var/data/src/efi
+ LD_LIBRARY_PATH=inst/usr/lib64
+ inst/usr/sbin/efibootmgr -c -L 'Debian (EFI stub) HEAD-u' -l '\EFI\debian\vmlinuz' -u 'root=UUID=a24f2065-ecd3-4f62-b402-a90f2e99cdcf rootfstype=btrfs  add_efi_memmap initrd=/EFI/debian/initrd.img'
+ efibootmgr -c -L 'Debian (EFI stub) 0.24-u' -l '\EFI\debian\vmlinuz1' -u 'root=UUID=a24f2065-ecd3-4f62-b402-a90f2e99cdcf rootfstype=btrfs  add_efi_memmap initrd=/EFI/debian/initrd.img'
+ LD_LIBRARY_PATH=inst/usr/lib64
+ inst/usr/sbin/efibootmgr -c -L 'Debian (EFI stub) HEAD' -l '\EFI\debian\vmlinuz2' 'root=UUID=a24f2065-ecd3-4f62-b402-a90f2e99cdcf rootfstype=btrfs  add_efi_memmap initrd=/EFI/debian/initrd.img'
+ efibootmgr -c -L 'Debian (EFI stub) 0.24' -l '\EFI\debian\vmlinuz3' 'root=UUID=a24f2065-ecd3-4f62-b402-a90f2e99cdcf rootfstype=btrfs  add_efi_memmap initrd=/EFI/debian/initrd.img'
run-parts: executing /etc/kernel/postinst.d/zz-update-grub 4.7.0-rc5-whirm+ /boot/vmlinuz-4.7.0-rc5-whirm+

└──> sudo efibootmgr -v
BootCurrent: 0001
Timeout: 0 seconds
BootOrder: 0003,0002,0001,0000
Boot0000* Debian (EFI stub) HEAD-u  HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\vmlinuz)r.o.o.t.=.U.U.I.D.=.a.2.4.f.2.0.6.5.-.e.c.d.3.-.4.f.6.2.-.b.4.0.2.-.a.9.0.f.2.e.9.9.c.d.c.f. .r.o.o.t.f.s.t.y.p.e.=.b.t.r.f.s. . .a.d.d._.e.f.i._.m.e.m.m.a.p. .i.n.i.t.r.d.=./.E.F.I./.d.e.b.i.a.n./.i.n.i.t.r.d...i.m.g.
Boot0001* Debian (EFI stub) 0.24-u  HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\vmlinuz1)r.o.o.t.=.U.U.I.D.=.a.2.4.f.2.0.6.5.-.e.c.d.3.-.4.f.6.2.-.b.4.0.2.-.a.9.0.f.2.e.9.9.c.d.c.f. .r.o.o.t.f.s.t.y.p.e.=.b.t.r.f.s. . .a.d.d._.e.f.i._.m.e.m.m.a.p. .i.n.i.t.r.d.=./.E.F.I./.d.e.b.i.a.n./.i.n.i.t.r.d...i.m.g.
Boot0002* Debian (EFI stub) HEAD    HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\vmlinuz2)root=UUID=a24f2065-ecd3-4f62-b402-a90f2e99cdcf rootfstype=btrfs  add_efi_memmap initrd=/EFI/debian/initrd.img
Boot0003* Debian (EFI stub) 0.24    HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\vmlinuz3)root=UUID=a24f2065-ecd3-4f62-b402-a90f2e99cdcf rootfstype=btrfs  add_efi_memmap initrd=/EFI/debian/initrd.img
```

All of them boot, but none of it gets the arguments passed to the kernel.

I can't add arguments when adding entries from the BIOS setup, so I can't compare it with anything.

**Comment by superm1** [comment]
You do want the extra arguments as UCS-2.
In the example that is here:
http://www.rodsbooks.com/efi-bootloaders/efistub.html, it's a relative path
passed for the initrd.
eg:

initrd=EFI/debian/initrd.img


Moot point though if the arguments still aren't getting passed.  Did you
build your kernel with CONFIG_CMDLINE_BOOL?  I think that will prevent it
taking any arguments passed from bootloader.

**Comment by whirm** [comment]
> You do want the extra arguments as UCS-2.

I guessed so, but as it wasn't working I tried both ways just in case.

> Moot point though if the arguments still aren't getting passed. Did you build your kernel with CONFIG_CMDLINE_BOOL? I think that will prevent it taking any arguments passed from bootloader.

I'm not setting `CONFIG_CMDLINE_OVERRIDE` which is the one that will make the kernel ignore the passed args according to the docs.

I'm guessing this is a BIOS bug? I've also tried packing the whole thing together with:

```
objcopy \
    --add-section .osrel=/etc/os-release --change-section-vma .osrel=0x20000 \
    --add-section .cmdline="cmdline.txt" --change-section-vma .cmdline=0x30000 \
    --add-section .linux="/boot/vmlinuz-4.7.0-rc4-amd64" --change-section-vma .linux=0x40000 \
    --add-section .initrd="/boot/initrd.img-4.7.0-rc4-amd64" --change-section-vma .initrd=0x3000000 \
    /usr/lib/systemd/boot/efi/linuxx64.efi.stub kernel.efi
```

And it's not receiving the arguments either (not from the EFI entry nor from cmdline.txt) The initramfs is found this way though.

**Comment by vathpela** [comment]
So, trying to figure out the -e 3 problem:
Boot0000* dobian    PciRoot(0x0)/Pci(0x17,0x0)/Sata(3,32768,0)/HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)
Boot0001* debian    PciRoot(0x0)/Pci(0x17,0x0)/Sata(2,65535,0)/HD(1,GPT,2929876f-fa70-476a-b8dd-ecd8def9a0dd,0x800,0x17579)/File(\EFI\debian\grubx64.efi)

From the UEFI Spec:
> Sata(HPN, PMPN, LUN)
> The HPN is an integer between 0 and 65534 and is required.
> The PMPN is an integer between 0 and 65535 and is optional. If not
> provided, the default is 0xFFFF, which implies no port multiplier.
> The LUN is a 16-bit integer. It is required. Note that LUN is applicable
> to ATAPI devices only, and most ATAPI devices assume LUN=0

So PMPN is the easy one - clearly I've got some type wrong and it's not converting right, since this literally says you're using a device connected to SATA Port Multiplier # 32767, whereas 65535 means no multiplier.  As PMPN is an enumeration done during traversal of devices, and you probably do *not* have 32k port multipliers attached to the machine, (and your linux device is /sys/class/ata_device/dev3.32767.0/) , this is us assigning some value with a wrong assumption about its storage type.

HPN also looks like off by one, probably based on firmware zero-indexing this value and (for reasons I that appear to be completely arbitrary) libata in the linux kernel using 1-based indices for this.  Let me give you some patches to try in a few minutes.

**Comment by whirm** [comment]
It looks right on my laptop too with the `sata` branch, thanks!

**Comment by vathpela** [comment]
This should be fixed in efibootmgr-13.  If there's still an issue, please reopen this, or reopen it as a separate issue if it's sufficiently different.

**Comment by concatime** [comment]
[Background: I have a Dell 7567, and faced pretty much the same issue]
Sorry for the necrobump, but for people stumbling across this, it seems that some (all?) UEFI implementations (firmware) from Dell does **[NOT](//bbs.archlinux.org/viewtopic.php?pid=1625106#p1625106)** aloud passing arguments…
A pragmatic solution would be relying on GRUB, but if you have another solution (beside `startup.nsh`, which I didn't try), I'm widely interested.
But the one thing that bothers me is why `-e3` is required to aloud a boot entry to show in boot menu? 
(It cannot boot from `HD`, only `PciRoot`)
